<a href="https://colab.research.google.com/github/unburied/DS-Unit-1-Sprint-3-Statistical-Tests-and-Experiments/blob/master/DS_Unit_1_Sprint_Challenge_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Science Unit 1 Sprint Challenge 3

## Exploring Data, Testing Hypotheses

In this sprint challenge you will look at a dataset of people being approved or rejected for credit.

https://archive.ics.uci.edu/ml/datasets/Credit+Approval

Data Set Information: This file concerns credit card applications. All attribute names and values have been changed to meaningless symbols to protect confidentiality of the data. This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

Attribute Information:
- A1: b, a.
- A2: continuous.
- A3: continuous.
- A4: u, y, l, t.
- A5: g, p, gg.
- A6: c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- A7: v, h, bb, j, n, z, dd, ff, o.
- A8: continuous.
- A9: t, f.
- A10: t, f.
- A11: continuous.
- A12: t, f.
- A13: g, p, s.
- A14: continuous.
- A15: continuous.
- A16: +,- (class attribute)

Yes, most of that doesn't mean anything. A16 (the class attribute) is the most interesting, as it separates the 307 approved cases from the 383 rejected cases. The remaining variables have been obfuscated for privacy - a challenge you may have to deal with in your data science career.

Sprint challenges are evaluated based on satisfactory completion of each part. It is suggested you work through it in order, getting each aspect reasonably working, before trying to deeply explore, iterate, or refine any given step. Once you get to the end, if you want to go back and improve things, go for it!

## Part 1 - Load and validate the data

- Load the data as a `pandas` data frame.
- Validate that it has the appropriate number of observations (you can check the raw file, and also read the dataset description from UCI).
- UCI says there should be missing data - check, and if necessary change the data so pandas recognizes it as na
- Make sure that the loaded features are of the types described above (continuous values should be treated as float), and correct as necessary

This is review, but skills that you'll use at the start of any data exploration. Further, you may have to do some investigation to figure out which file to load from - that is part of the puzzle.

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [9]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data

--2019-06-03 01:20:24--  https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 32218 (31K) [application/x-httpd-php]
Saving to: ‘crx.data’

crx.data            100%[===================>]  31.46K  --.-KB/s    in 0.07s   

2019-06-03 01:20:25 (461 KB/s) - ‘crx.data’ saved [32218/32218]



In [10]:
!ls

Credit+Approval  crx.data  sample_data


In [12]:
!head -5 crx.data

b,30.83,0,u,g,w,v,1.25,t,t,01,f,g,00202,0,+
a,58.67,4.46,u,g,q,h,3.04,t,t,06,f,g,00043,560,+
a,24.50,0.5,u,g,q,h,1.5,t,f,0,f,g,00280,824,+
b,27.83,1.54,u,g,w,v,3.75,t,t,05,t,g,00100,3,+
b,20.17,5.625,u,g,w,v,1.71,t,f,0,f,s,00120,0,+


In [0]:
columns = [f'A{i}' for i in range(1, 17)]

df = pd.read_csv('crx.data', names = columns)

In [17]:
df.head()

A1     A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +

In [19]:
df.shape

(690, 16)

In [20]:
df.describe()

A3          A8        A11            A15
count  690.000000  690.000000  690.00000     690.000000
mean     4.758725    2.223406    2.40000    1017.385507
std      4.978163    3.346513    4.86294    5210.102598
min      0.000000    0.000000    0.00000       0.000000
25%      1.000000    0.165000    0.00000       0.000000
50%      2.750000    1.000000    0.00000       5.000000
75%      7.207500    2.625000    3.00000     395.500000
max     28.000000   28.500000   67.00000  100000.000000

In [22]:
df.dtypes

A1      object
A2      object
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11      int64
A12     object
A13     object
A14     object
A15      int64
A16     object
dtype: object

In [27]:
#replace all ? with NA values
#in hindsight couldve used df.replace 

for col in columns:
  df[col] = np.where(df[col] == '?', np.NaN, df[col])

/usr/local/lib/python3.6/dist-packages/pandas/core/ops.py:1649: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


In [0]:
#corrected data types per documentation 

df['A2'] = df.A2.astype('float')
df['A14'] = df.A14.astype('float')

In [31]:
df.dtypes

A1      object
A2     float64
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11    float64
A12     object
A13     object
A14    float64
A15    float64
A16     object
dtype: object

In [33]:
df.isna().sum()

A1     12
A2     12
A3      0
A4      6
A5      6
A6      9
A7      9
A8      0
A9      0
A10     0
A11     0
A12     0
A13     0
A14    13
A15     0
A16     0
dtype: int64

## Part 2 - Exploring data, Testing hypotheses

The only thing we really know about this data is that A16 is the class label. Besides that, we have 6 continuous (float) features and 9 categorical features.

Explore the data: you can use whatever approach (tables, utility functions, visualizations) to get an impression of the distributions and relationships of the variables. In general, your goal is to understand how the features are different when grouped by the two class labels (`+` and `-`).

For the 6 continuous features, how are they different when split between the two class labels? Choose two features to run t-tests (again split by class label) - specifically, select one feature that is *extremely* different between the classes, and another feature that is notably less different (though perhaps still "statistically significantly" different). You may have to explore more than two features to do this.

For the categorical features, explore by creating "cross tabs" (aka [contingency tables](https://en.wikipedia.org/wiki/Contingency_table)) between them and the class label, and apply the Chi-squared test to them. [pandas.crosstab](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html) can create contingency tables, and [scipy.stats.chi2_contingency](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) can calculate the Chi-squared statistic for them.

There are 9 categorical features - as with the t-test, try to find one where the Chi-squared test returns an extreme result (rejecting the null that the data are independent), and one where it is less extreme.

**NOTE** - "less extreme" just means smaller test statistic/larger p-value. Even the least extreme differences may be strongly statistically significant.

Your *main* goal is the hypothesis tests, so don't spend too much time on the exploration/visualization piece. That is just a means to an end - use simple visualizations, such as boxplots or a scatter matrix (both built in to pandas), to get a feel for the overall distribution of the variables.

This is challenging, so manage your time and aim for a baseline of at least running two t-tests and two Chi-squared tests before polishing. And don't forget to answer the questions in part 3, even if your results in this part aren't what you want them to be.

In [0]:
from scipy import stats

In [0]:
approved = df[df.A16 == '+']
denied = df[df.A16 == '-']

In [0]:
sns.pairplot(approved, kind = 'reg');

In [0]:
sns.pairplot(denied, kind = 'reg');

In [0]:
approved.describe()

In [0]:
denied.describe()

In [57]:
stats.ttest_ind(approved.A11, denied.A11, equal_var = False, nan_policy = 'omit')
#Reject null hypothesis that the mean of Approved column A11 is equal to Denied column A11
#most extreme

Ttest_indResult(statistic=10.6384190682749, pvalue=4.310254123415665e-23)

In [53]:
stats.ttest_ind(approved.A14, denied.A14, equal_var = False, nan_policy = 'omit')
#Reject null hypothesis that the mean of Approved column A14 is equal to Denied column A14
#least extreme

Ttest_indResult(statistic=-2.6696493137915973, pvalue=0.0077778250827957)

In [0]:
# test features using chi_squared and determine most and least significant 
def chi(features):
  chi_stats = {}
  
  #list to use object columns
  cols = [col for col in features if df[col].dtype == 'O']
  
  #initiate chi test for each column and store in dictionary
  for column in cols:
    table = pd.crosstab(df[column], df.A16) #compare all to class labels
    chi, pval, _,_ = stats.chi2_contingency(table)
    chi_stats[column] = [chi,pval]
  
  #del test for A16 against itself
  del chi_stats['A16']
  
  return chi_stats
  
  

In [70]:
tests = chi(columns)
tests

{'A1': [0.31128326491619945, 0.5768937883001117],
 'A10': [143.06956205083145, 5.675727374527571e-33],
 'A12': [0.568273300792113, 0.45094587758631943],
 'A13': [9.191570451545383, 0.010094291370456362],
 'A4': [26.234074966202144, 2.010680204180363e-06],
 'A5': [26.234074966202144, 2.010680204180363e-06],
 'A6': [98.32520342679135, 3.4999300402715717e-15],
 'A7': [45.03420714024056, 3.62545287237226e-07],
 'A9': [355.2038167412799, 3.1185900878457007e-79]}

## Part 3 - Analysis and Interpretation

Now that you've looked at the data, answer the following questions:

- Interpret and explain the two t-tests you ran - what do they tell you about the relationships between the continuous features you selected and the class labels?
- Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?
- What was the most challenging part of this sprint challenge?

Answer with text, but feel free to intersperse example code/results or refer to it from earlier.

####Interpret and explain the two t-tests you ran - what do they tell you about the relationships ####between the continuous features you selected and the class labels?

Ttest_indResult(statistic=10.6384190682749, pvalue=4.310254123415665e-23)
Reject null hypothesis that the mean of Approved column A11 is equal to Denied column A11

This was the most extreme example due to the means being so significantly different. Not only were the means different, but the variance in the approved set was several times larger than the variance of the denied set. Due to the smaller (but not smallest) totals and the large variance, I'm going to infer this columns represents the number of Trade Lines on a credit report. The relationship to the class labels is significant.

Ttest_indResult(statistic=-2.6696493137915973, pvalue=0.0077778250827957)
Reject null hypothesis that the mean of Approved column A14 is equal to Denied column A14

This was the least extreme comparison made among continuous variables involved. This was also one of the only features where the approved set had lower statistical values overall compared to the denied set. I would infer this has something to do with credit utilization. Most likely it is associated with the monthly payments due on current credit accounts. Applicants with better credit tend to pay off debt quicker, which leads to lower monthly payments. The relationship to the class labels is less significant, although still essential. 



####Interpret and explain the two Chi-squared tests you ran - what do they tell you about the relationships between the categorical features you selected and the class labels?

{'A1': [0.31128326491619945, 0.5768937883001117],

 'A10': [143.06956205083145, 5.675727374527571e-33],
 
 'A12': [0.568273300792113, 0.45094587758631943],
 
 'A13': [9.191570451545383, 0.010094291370456362],
 
 'A4': [26.234074966202144, 2.010680204180363e-06],
 
 'A5': [26.234074966202144, 2.010680204180363e-06],
 
 'A6': [98.32520342679135, 3.4999300402715717e-15],
 
 'A7': [45.03420714024056, 3.62545287237226e-07],
 
 'A9': [355.2038167412799, 3.1185900878457007e-79]}
 
 I thought it was curious that there were 4 binomial features. Of those 4, half were statistically dependant with relation to the class labels.The other 2 features (A1, A12) were the only features were we failed to reject the null hypothesis. Even more interesting is that 3 out of those 4 had the same identifiers 't' and 'f', but 1(A12) was not anywhere near showing significant signs of dependance. All other features showed some level of dependancy with relation to the class labels, with the most significant being A9, one of the four mentioned above. One of the least extreme yet still significant features was A13 with a p-value of 1%.

####What was the most challenging part of this sprint challenge?

The most difficult aspect of this sprint challenge was controlling the urge to explore the data more. I wanted to find concrete expamples and 'solve' the code to find what was hidden. I had to balance that urge with the need to manage my time wisely to ensure I completed each and every task. I did feel a little more comfortable with the concepts involved, but there was still a feeling that I hadn't done the tasks correctly or completely. 